src/DIRAC/WorkloadManagementSystem/Agent/JobAgent.py

```diff
@@ -426,37 +329,120 @@ def __saveJobJDLRequest(self, jobID, jobJDL):
         jdlFile.close()
 
     #############################################################################
-    def computeCPUWorkLeft(self, processors=1):
+    def _getCEDict(self, computingElement):
+        """Get CE description
+
+        :param ComputingElement computingElement: ComputingElement instance
+        :return: list of dict of attributes
+        """
+        # if we are here we assume that a job can be matched
+        result = computingElement.getDescription()
+        if not result["OK"]:
+            self.log.warn("Can not get the CE description")
+            return result
+
+        # We can have several prioritized job retrieval strategies
+        if isinstance(result["Value"], dict):
+            ceDictList = [result["Value"]]
+        else:
+            # This is the case for Pool ComputingElement, and parameter 'MultiProcessorStrategy'
+            ceDictList = result["Value"]
+
+        return S_OK(ceDictList)
+
+    def _setCEDict(self, ceDict):
+        """Set CEDict"""
+        # Add pilot information
+        gridCE = gConfig.getValue("LocalSite/GridCE", "Unknown")
+        if gridCE != "Unknown":
+            ceDict["GridCE"] = gridCE
+        if "PilotReference" not in ceDict:
+            ceDict["PilotReference"] = str(self.pilotReference)
+        ceDict["PilotBenchmark"] = self.cpuFactor
+        ceDict["PilotInfoReportedFlag"] = self.pilotInfoReportedFlag
+
+        # Add possible job requirements
+        result = gConfig.getOptionsDict("/AgentJobRequirements")
+        if result["OK"]:
+            requirementsDict = result["Value"]
+            ceDict.update(requirementsDict)
+            self.log.info("Requirements:", requirementsDict)
+
+    def _checkCEAvailability(self, computingElement):
+        """Check availability of computingElement"""
+        result = computingElement.available()
```

Comment:
A minor comment about `checkCEAvailability`:
the method uses `ComputingElement.available()`, which uses `ComputingElement.getCEStatus()` to know whether a given CE can get further jobs.

CE interfaces such as `ARCCE` can interrogate the CE to get information about running and waiting jobs.
However, interfaces such as `HTCondorCE` will return information from the `PilotAgentsDB` which will be wrong in this case.
Not really an issue now as we plan to use it mostly with `ARC` but might become one in the future...

See: https://github.com/DIRACGrid/DIRAC/pull/5406